docs/DeveloperGuide.adoc

```diff
@@ -767,20 +817,258 @@ Use case ends.
+Error: AddressBook informs user that input is invalid.
+
+[discrete]
+=== Use case: List Contact
+
+
+*MSS*
+
+1.  User requests to list persons
+2.  AddressBook shows a list of persons
+
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+
+
+[discrete]
+=== Use case: Edit Contact
+
+
+*MSS*
+
+1.  User requests to edit person (index) and specific particular
+2.  AddressBook edits the contact particular and displays latest change
++
+Use case ends.
+
+*Extensions*
+
+[none]
+
+*3a. User requests to update existing contact.
+*3b. User selects from existing contact(s).
+*3c. AddressBook updates selected contact.
++
+Use case ends.
++
+**3b1. The given index is invalid.
++
+***3b11. AddressBook shows an error message.
++
+Use case resumes at step 3b.
+
+[discrete]
+=== Use case: Clear
+
+*Preconditions*
+1. AddressBook is not empty.
+
+*MSS*
+
+1. User requests to clear list.
+2. AddressBook clears list.
++
+Use case ends.
+
+[discrete]
+=== Use case: Exit
+
+*MSS*
+
+1. User requests to exit the program.
+2. AddressBook closes the program.
++
+Use case ends.
+
+[discrete]
+=== Use case: History
+
+*MSS*
+
+1. User requests history of commands issued.
+2. AddressBook shows list of commands in reverse chronological order.
++
+Use case ends.
+* 1a. Person Index is invalid.
++
+AddressBook informs user that Person Index is invalid.
+
+* 1b. User requests to edit information in an invalid format
++
+AddressBook informs user that the format is invalid, providing the correct format.
+
+
+[discrete]
+=== Use case: Find Contact
+
+
+*MSS*
+
+
+ 1.  User requests to find person by entering part of contact name
+ 2.  AddressBook finds contact with keyword included in name, and displays latest change
+ +
+ Use case ends.
+
+*Extensions*
+
+[none]
+* 1a. Contact cannot be found
++
+AddressBook informs user that contact cannot be found
+
+[discrete]
+=== Use case: Select Contact
+
+
+*MSS*
+
+1.  User requests to see entire list of contacts
+2.  Based on list, user selects contact by providing contact index.
+3.  AddressBook selects contact with index given.
+
++ Use case ends.
+
+*Extensions*
+
+[none]
+*2a. The list is empty.
++
+Use case ends.
+
+* 1a. No contacts in list.
++
+AddressBook informs user that there are currently no contacts in the list.
+
+* 2a. User inputs invalid contact index.
++
+AddressBook informs user that the input is not a valid contact index.
 
 [appendix]
 == Non Functional Requirements
 
 .  Should work on any link:#mainstream-os[mainstream OS] as long as it has Java `1.8.0_60` or higher installed.
 .  Should be able to hold up to 1000 persons without a noticeable sluggishness in performance for typical usage.
 .  A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse.
+.  Data held in addressbook cannot be corrupted in event of unexpected errors.
```

Comment:
This sounds good on the surface, however, this is almost impossible to achieve, as there's no definition of "unexpected errors" in the context of this NFR. "Unexpected errors" could mean that the hard drive got corrupted and the database file loses its integrity.